api/logical.go

```diff
@@ -97,6 +97,9 @@ func (c *Logical) ReadWithDataWithContext(ctx context.Context, path string, data
 		if secret != nil && (len(secret.Warnings) > 0 || len(secret.Data) > 0) {
 			return secret, nil
 		}
+		if c.c.ReturnErrorOnMissing() {
+			return nil, err
+		}
```

Comment:
Is it possible that the error returned from `rawRequestWithContext` is nil when `StatusCode == 404`? If so, would this will also return `nil, nil`? 

I'm also not sure if we need to call `ParseSecret` at all when `ReturnErrorOnMissing()` is true.

Reply:
It is possible for err to be nil, and the `StatusCode == 404`. Wondering what would be the concern? Would you please elaborate more? I think all we want to do here is to return err, but the err could also be nil. 
An example of a place where err is nil and code is 404: https://github.com/hashicorp/vault/blob/fcbe948b2542a13ee8036ad07dd8ebf8554f56cb/http/handler.go#L366

I looked at almost all of read operations. The majority of times, if the resource is not found, `nil, nil` (resp, err) is returned. These cases normally is handled here:
https://github.com/hashicorp/vault/blob/fcbe948b2542a13ee8036ad07dd8ebf8554f56cb/sdk/logical/response_util.go#L22

For parsing a secret, the only parameters in the response that is of interest in this version of the api are Data and Warnings. For those two, parseSecret is already called. If anything is there, line 97 would result in true, and we would return the secret. 

I say this version of the api because there is another field `WrapInfo` in the [api.Secret](https://github.com/hashicorp/vault/blob/fcbe948b2542a13ee8036ad07dd8ebf8554f56cb/api/secret.go#L39) which will be populated when control groups are involved. But, the current version of the api is not FULLY able to handle it. I say fully cause there are places where it is handled. Changing it would be a breaking change, backwards incompatibility nightmare. There is bug that has more info on it [here](https://hashicorp.atlassian.net/browse/VAULT-2140). So, I am not sure if there is anything else interesting in the resp to grant us call `parseSecret` when `ReturnErrorOnMissing` is true.

Reply:
Sorry for the late response. My concern was about the user expectations when they set `VAULT_RETURN_ERROR_ON_MISSING` or `ReturnErrorOnMissing()` flag and yet still get back `nil, nil` from the method call. This might be a surprising behavior for some.

What I was thinking to work around this issue is to always return an error when the flag is set and only parse the secret after the check:

```go
if resp != nil && resp.StatusCode == 404 {
	if c.c.ReturnErrorOnMissing() {
		return nil, fmt.Errorf("secret not found at %q: %w", path, err) // not sure if we should include the original `err` here or not
	}

	secret, parseErr := ParseSecret(resp.Body)
	...
}
```